### PR TITLE
Send before adds message expiration

### PR DESCRIPTION
**Pull request by sduckett**
This PR adds an optional `:send-before` key that we can use to expire messages down the road. Once downstream producers of messages are adding the key, we'll make it a required key and come back to scheduled-message-works

Testing with a production-style krakenstein, all of our notification types can be created, scheduled, sent, and removed. It's a bit of a tricky set of diffs to read, so please ask questions.

[Pivotal Card](https://www.pivotaltracker.com/story/show/115346585)

### CODE REVIEW

File: src/scheduled_message_works/entities/message.clj [hunk 1]
```diff
@@ -141,8 +170,9 @@
 (defn update-message [existing-message new-message]
   (s/validate schedule-message-schema new-message)
   (let [existing-message-id (:scheduled-message/id existing-message)
-        tx-data (->tx-data (:db/id existing-message)
-                           (assoc new-message :id existing-message-id))
+        tx-data (map #(tx/dissoc-optional-nil-values % [:scheduled-message/send-before])
```
Comment (cap10morgan) [inline src/scheduled_message_works/entities/message.clj]: Dissoc'ing optional nil values should be moved into the `->tx-data` function. There are a few examples of this scattered around our -works components.

File: src/scheduled_message_works/scheduler.clj [hunk 2]
```diff
@@ -32,12 +32,23 @@
   and sends them."
   [sent-at-before]
   (log/info "sending scheduled messages from before" sent-at-before)
-  (if-let [ready-messages (message/get-ready-messages
+  (if-let [ready-messages (message/sendable-messages
```
Comment (cap10morgan) [inline src/scheduled_message_works/scheduler.clj]: Let's bind to `sendable-messages` or similar here if we're changing the name of the fn.

File: test/scheduled_message_works/entities/message_test.clj [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        send-at (time/plus start (time/days 1))
+        send-before (time/plus start (time/days 2))
+
+        complete-message (td/test-message-entity (coerce/to-date send-at)
+                                                 (coerce/to-date send-before))
+        complete-result (message/schedule-message complete-message)]
+
+    (time/do-at
+     start
+     (testing "when both `:send-before` is future, nothing is stale"
```
Comment (cap10morgan) [inline test/scheduled_message_works/entities/message_test.clj]: I don't understand the test description here.

File: src/scheduled_message_works/scheduler.clj [hunk 4]
```diff
@@ -32,12 +32,23 @@
...
+(defn remove-stale-messages
+  "Remove the set of messages that have an expiration before `time`, logging
+  what we did."
+  [time]
+  (if-let [stale-messages (message/stale-messages
+                           (db/db)
+                           (coerce/to-date time))]
+    (doseq [m (map message/rehydrate stale-messages)]
+      (message/delete-message (:id m))
+      (log/info "removed stale message: " (pr-str m)))))
```
Comment (sduckett) [inline src/scheduled_message_works/scheduler.clj]: :+1:

File: test/scheduled_message_works/entities/message_test.clj [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                (coerce/to-date (time/now))))))))
+
+    (time/do-at
+     (time/plus start (time/days 2) (time/seconds 1))
+     (testing "when time passes `:send-before` a message becomes stale"
+       (is (= 1
+              (count
+               (message/stale-messages
+                (dt/db)
+                (coerce/to-date (time/now))))))))))
```
Comment (cap10morgan) [inline test/scheduled_message_works/entities/message_test.clj]: Yay for tests!

File: src/scheduled_message_works/scheduler.clj (hunk 2 above)
Comment (sduckett) [inline src/scheduled_message_works/scheduler.clj]: :+1:

File: test/scheduled_message_works/entities/message_test.clj (hunk 3 above)
Comment (sduckett) [inline test/scheduled_message_works/entities/message_test.clj]: Nor do I...I think it's an incomplete edit.

File: src/scheduled_message_works/entities/message.clj (hunk 1 above)
Comment (sduckett) [inline src/scheduled_message_works/entities/message.clj]: I'll have a look. `->tx-data` is the child of a macro here....

Comment (cap10morgan) [inline src/scheduled_message_works/entities/message.clj]: A macro I'm not sure is useful enough to justify its existence. There are some non-macro versions of `->tx-data` floating around (let me know if you can't find one in a -works component) that I think I prefer.

**Comment by cap10morgan** [comment]
LGTM

**Pull request merged**